File: 2882-ways-to-express-an-integer-as-sum-of-powers/ways-to-express-an-integer-as-sum-of-powers.js
/**
 * @param {number} n
 * @param {number} x
 * @return {number}
 */
var numberOfWays = function (n, x) {
    const dp = Array(n + 1).fill(0);
    const mod = 10 ** 9 + 7;

    dp[0] = 1;

    for (let i = 1; i < n + 1; i++) {
        let m = i ** x;
        if (m > n) {
            break;
        }
        for (let j = n; j >= m; j--) {
            dp[j] += dp[j - m];
        }
    }

    return dp[n] % mod;
};

File: 2882-ways-to-express-an-integer-as-sum-of-powers/ways-to-express-an-integer-as-sum-of-powers.py
class Solution:
    def numberOfWays(self, n: int, x: int) -> int:
        dp = [0] * (n + 1)
        mod = 10**9 + 7

        dp[0] = 1

        for num in range(1, n + 1):
            m = num**x
            for i in range(n, 0, -1):
                c = i - m
                if c >= 0:
                    dp[i] += dp[c]
                    dp[i] %= mod

        return dp[n]
